smb: add smb.version keyword

jmtaylor90 opened the pull request:
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

https://redmine.openinfosecfoundation.org/issues/5075

Describe changes:
- adds smb.version keyword
- adds smb.version documentation
-

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_BRANCH=pr/1366
```

catenacyber commented:
Did you start from https://github.com/OISF/suricata/pull/7336 ?

catenacyber commented:
Check rust failure seems unrelated : fix may be https://github.com/OISF/suricata/pull/9416/commits

jmtaylor90 commented:
> Did you start from #7336 ?

I did, sorry about that, I was going to reference/indicate that and forgot.

jmtaylor90 commented:
> > > Did you start from #7336 ?
> > 
> > 
> > I did, sorry about that, I was going to reference/indicate that and forgot.
> 
> Thanks for picking this back up. What did you change ? (just rebase ?)

This was largely a rebase and formatting (clang/clippy/rustfmt) changes. I did add some debug logging as well.

I did change some code to address an initial question you had around:
https://github.com/OISF/suricata/pull/7336/files#diff-f037023d6eda2c0a79e08ad1701daa9cd8842be75ba08c267990edc50d51926fR98

I removed that bit of code and updated the DetectAppLayerInspectEngineRegister2 calls.

I also added an additional signature validation check:
https://github.com/OISF/suricata/pull/9415/files#diff-f037023d6eda2c0a79e08ad1701daa9cd8842be75ba08c267990edc50d51926fR90

which is verified in the associated suricata verify  PR

catenacyber commented on src/detect-smb-version.h:
```diff
@@ -0,0 +1,25 @@
+ *         detect.c::SigTableSetup() */
+void DetectSmbVersionRegister(void);
+
+#endif /* __DETECT_SMB-VERSION_H__ */
```
typo - _

jmtaylor90 replied:
fixed in next PR

catenacyber commented on src/detect-smb-version.c:
```diff
@@ -0,0 +1,165 @@
+
+#define BUFFER_NAME         "smb_version"
+#define KEYWORD_NAME        "smb.version"
+#define KEYWORD_NAME_LEGACY BUFFER_NAME
```
no legacy needed

jmtaylor90 replied:
fixed in next PR

catenacyber reviewed (changes requested):
Thanks for the changes

- CI : ✅  rust failure is unrelated
- Code : looks good, some nits to review
- Commits segmentation : maybe the rustfmt can be its own first commit
- Commit messages : Could you put the ticket number in the commit message please ? 
- Git ID set : looks fine for me
- CLA : I do not have access, but looks like some commits were already merged
- Doc update : really nice  👏 
- Redmine ticket : ok
- Rustfmt : ok
- Tests : Suricata-verify tests look fine
- Dependencies added: none

jmtaylor90 commented on src/detect-smb-version.c:
```diff
@@ -0,0 +1,165 @@
+        return rs_smb_version_match(txv, (void *)m);
+    }
+
+    if (rs_smb_version_match(txv, (void *)m) != 1)
```
I am not sure, I saw this sort of if block pattern in a couple other places and figured it came in here in matching style. Example : https://github.com/OISF/suricata/blob/master/src/detect-dce-opnum.c#L105

I also came across:
https://github.com/OISF/suricata/blob/master/src/detect-dce-iface.c#L114

Where there is some logging added. Looking at this again, I kind of like the additional logging but of course defer to what you think the best route to go is here.

Additional logging wise I was thinking something like:

```
if (f->alproto == ALPROTO_SMB) {
        return rs_smb_version_match(txv, (void *)m);
    }

    if (rs_smb_version_match(txv, (void *)m) != 1) {
        SCLogDebug("rs_smb_version_match: didn't match");
        SCReturnInt(0);
    } else {
        SCLogDebug("rs_smb_version_match: matched!");
    }

    SCReturnInt(1);
}
```

I think ALPROTO_SMB is the right one to use here for obtaining version information.

catenacyber replied:
So, I think DCERPC keywords can work for both SMB and DCERPC, but SMB keywords are SMB only, as enforced in the setup function.
So you need to call only once `rs_smb_version_match` and not check for `f->alproto == ALPROTO_SMB`

jmtaylor90 replied:
Ah okay, I see. I will see about removing that check. What do you think with regards to adding the logging block?

jufajardini commented on rust/src/smb/detect.rs:
```diff
@@ -156,11 +152,9 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
-{
+pub unsafe extern "C" fn rs_smb_tx_get_ntlmssp_user(
+    tx: &mut SMBTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
```
Are these purely formatting changes to existing code? If yes, we usually prefer to see those in a separate commit, so its easier to visualize and review :)

jmtaylor90 replied:
They are, I will have the rustfmt/clippy changes in a separate commit in the next PR.

jufajardini replied:
Happens to us all, thanks! :)

jufajardini commented on doc/userguide/rules/smb-keywords.rst:
```diff
@@ -58,3 +58,45 @@ Examples::
+
+The Protocol ID in header distinguishes only SMB1 and SMB2 since they are totally different protocols with total different message formats, types and implementation.
+
+On the other hand SMB3 is more an extension for SMB2. When using SMB2 we can select one of the following dialects for the conversation between client and server:
+
+- 2.0.2
+- 2.1
+- 3.0
+- 3.0.2
+- 3.1.1
+
+We say we are using SMB3 when we select a 3.x dialect for the conversation, so you can use SMB3.0, SMB3.0.2 or SMB3.1.1. The higher you choose, the more capabilities you have, but the message syntax and message command number remains the same.
```
While I liked the more in-depth explanation here, it made me a bit confused as to how this impacts the keyword usage, if at all. 

Can you add a note on this, or maybe complement it with rule examples, if this is something that is relevant to rule writers? :)

jmtaylor90 replied:
Sure, will revisit for the next PR and update the -verify tests as needed

jufajardini commented on rust/src/smb/detect.rs:
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2017 Open Information Security Foundation
+/* Copyright (C) 2023 Open Information Security Foundation
```
Copyright years are updated in the form of: initial year-last updated year
See https://github.com/OISF/suricata/pull/9415/commits/482eb11213677d4f2b56e52757b5810b112cd67d#diff-809d44481dd655fd192c9767d520c5ced78d5c1de6f2b94ee4255e91c1376d53L1 ;)

jmtaylor90 replied:
updated in next PR

jmtaylor90 commented:
continued in #9427
